core/server/worker/src/main/java/alluxio/worker/block/management/TierManagementTaskProvider.java

```diff
@@ -0,0 +1,133 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block.management;
+
+import alluxio.collections.Pair;
+import alluxio.conf.PropertyKey;
+import alluxio.conf.ServerConfiguration;
+import alluxio.worker.block.BlockMetadataEvictorView;
+import alluxio.worker.block.BlockMetadataManager;
+import alluxio.worker.block.BlockStore;
+import alluxio.worker.block.BlockStoreLocation;
+import alluxio.worker.block.meta.StorageTier;
+import alluxio.worker.block.annotator.BlockOrder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.function.Supplier;
+
+/**
+ * {@link ManagementTaskProvider} implementation for tier management tasks.
+ *
+ * It currently creates two types of tasks:
+ *  1- TierSwap task for swapping blocks between tiers in order to promote/demote blocks.
+ *  2- TierMove task for utilizing speed of higher tiers by moving blocks from below.
+ */
+public class TierManagementTaskProvider implements ManagementTaskProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(TierManagementTaskProvider.class);
+
+  private final BlockStore mBlockStore;
+  private final BlockMetadataManager mMetadataManager;
+  private final Supplier<BlockMetadataEvictorView> mEvictorViewSupplier;
+  private final StoreLoadTracker mLoadTracker;
+
+  /**
+   * Creates a task provider for tier management functions.
+   *
+   * @param blockStore block store
+   * @param metadataManager meta manager
+   * @param evictorViewSupplier evictor view supplier
+   * @param loadTracker load tracker
+   */
+  public TierManagementTaskProvider(BlockStore blockStore, BlockMetadataManager metadataManager,
+      Supplier<BlockMetadataEvictorView> evictorViewSupplier, StoreLoadTracker loadTracker) {
+    mBlockStore = blockStore;
+    mMetadataManager = metadataManager;
+    mEvictorViewSupplier = evictorViewSupplier;
+    mLoadTracker = loadTracker;
+  }
+
+  @Override
+  public BlockManagementTask getTask() {
+    switch (findNextTask()) {
+      case NONE:
+        return null;
+      case TIER_SWAP:
+        return new TierSwapTask(mBlockStore, mMetadataManager, mEvictorViewSupplier.get(),
+            mLoadTracker);
+      case TIER_MOVE:
+        return new TierMoveTask(mBlockStore, mMetadataManager, mEvictorViewSupplier.get(),
+            mLoadTracker);
+      default:
+        throw new IllegalArgumentException("Unknown task type.");
+    }
+  }
+
+  /**
+   * @return the next tier management task that is required
+   */
+  private TierManagementTaskType findNextTask() {
```

Comment:
After this method, we only hint at what kind of task is necessary but not the details of the task. What is the computation cost here and could we do more computation to pass onto the execution of the task (ie. which blocks to swap)? Also since the swap will happen at a different time, are there any consistency issues?

Reply:
Consistency is loose based on `ConcurrentSkipListSet`'s iterator consistency. So decision given by this method, or the later swap-list that is acquired could no longer be valid. In that case, next run is supposed to fix that. However, with load tracking, inconsistency will not be vast and will be sorted out on next run.

As for the computations, this method's heaviest lift is the `BlockIterator:overlaps` method which is not directly translatable. I wouldn't worry about the wasted work here as overlap will be detected at first block comparison at a tier intersection.